pkg/operator/util/ca.go

```diff
@@ -0,0 +1,71 @@
+package util
+
+import (
+	crand "crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"time"
+
+	cryptutil "github.com/openshift/library-go/pkg/crypto"
+)
+
+const SigningCertificateLifetimeInDays = 365 // 1 year
+
+// CertHalfwayExpired indicates whether half of the cert validity period has elapsed.
+func CertHalfwayExpired(cert *x509.Certificate) bool {
+	halfValidPeriod := cert.NotAfter.Sub(cert.NotBefore).Nanoseconds() / 2
+	halfExpiration := cert.NotBefore.Add(time.Duration(halfValidPeriod) * time.Nanosecond)
+	return time.Now().After(halfExpiration)
+}
+
+// RotateSigningCA creates a new CA with a set of cross-signed interim CAs to allow for graceful rollover.
+// Returns new CA pem, new CA key pem and the full client CA bundle.
+func RotateSigningCA(currentCert *x509.Certificate, currentKey *rsa.PrivateKey) ([]byte, []byte, []byte, error) {
+	// Generate a new signing cert
+	newCertConfig, err := cryptutil.MakeSelfSignedCAConfigForSubject(currentCert.Subject, SigningCertificateLifetimeInDays)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	newCert := newCertConfig.Certs[0]
+	newKey := newCertConfig.Key
+
+	// The first interim CA comprises the current CA's public key, private key, and subject. It's self-issued but not
+	// self-signed as it's signed by the new CA key. This creates a trust bridge between refreshed clients and
+	// unrefreshed servers.
+	rawSignedByNewCert, err := x509.CreateCertificate(crand.Reader, currentCert, currentCert, currentCert.PublicKey, newKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	signedByNewCert := &x509.Certificate{Raw: rawSignedByNewCert}
+
+	// The second interim CA comprises the new CA's public key, private key, and subject. It's self-issued but not
+	// self-signed as it's signed by the current CA key. This creates a trust bridge between the unrefreshed clients and
+	// refreshed servers as long as refreshed servers serve with a bundle containing this CA and the serving cert.
+	rawSignedByCurrentCert, err := x509.CreateCertificate(crand.Reader, newCert, newCert, newCert.PublicKey, currentKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	signedByCurrentCert := &x509.Certificate{Raw: rawSignedByCurrentCert}
+
+	// Create the full CA bundle comprising all 4 certs
+	// TODO(marun) Is the ordering of certs in the bundle important?
+	bundleCerts := []*x509.Certificate{
+		currentCert,
+		signedByNewCert,
+		signedByCurrentCert,
+		newCert,
+	}
+	bundle, err := cryptutil.EncodeCertificates(bundleCerts...)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// TODO(marun) Is it necessary to append signedByCurrentCert to
+	// newCertConfig.Certs? PR #42 was doing so.
```

Comment:
See the above comment, I drawed this on our whiteboard and if I am correct, we wouldn't need the `signedByNewCert` instead

Reply:
I've come to the conclusion that this was never necessary (not sure why #42 was doing so) - CA cert `#4` needs to be included with each new service cert rather than being included with the signing CA.

Reply:
Re-reading the doc and thinking about it once again, that does sound like a correct assumption given the #4 is distributed by the server and #3 will never really sign anything.

Reply:
Going back to this comment, how do you imagine that `#4` is distributed by the server if it is not appended to generated serving certs?